File: leetcode/STR/20.ValidParetheses/20ValidParetheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mark = []
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '{' or s[i] == '[':
                mark.append(s[i])
            elif s[i] == ')' or s[i] == '}' or s[i] == ']':
                if len(mark) == 0:
                    return False
                cur = mark.pop()
                if cur == '(' and s[i] != ')':
                    return False
                if cur == '{' and s[i] != '}':
                    return False
                if cur == '[' and s[i] != ']':
                    return False
        if len(mark) == 0:
            return True
        return False

File: leetcode/STR/20.ValidParetheses/20ValidParetheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function(s) {
  let result = true;
  const left = ["(", "[", "{"];
  const right = [")", "]", "}"];
  const compare = [];
  Array.from(s).forEach((item, idx, array) => {
    if (left.includes(item)) {
      compare.push(right[left.indexOf(item)]);
    } else {
      let curr = compare.pop();
      if (curr !== item) {
        result = false;
      }
    }
  });
  if (compare.length > 0) {
    return false;
  }
  return result;
};
